tools/unit/src/main/kotlin/com/datadog/tools/unit/ObjectTest.kt

```diff
@@ -0,0 +1,204 @@
+        val y = createInstance(forge)
+        assumeTrue(x !== y) { "Can't check equals transitivity: x:$x is same instance as y:$y" }
+
+        // When
+        val resultA = x.equals(y)
+        val resultB = y.equals(x)
+
+        // Then
+        assertThat(resultA).isEqualTo(resultB)
+    }
+
+    /**
+     * `equals` must be transitive.
+     * if (x.equals(y) && y.equals(z)) then x.equals(z)
+     */
+    @Test
+    fun `equals is transitive`(forge: Forge) {
+        // Given
+        val x = createInstance(forge)
+        val y = createEqualInstance(x, forge)
+        val z = createEqualInstance(y, forge)
+        assumeTrue(x !== y) { "Can't check equals transitivity: x:$x is same instance as y:$y" }
+        assumeTrue(y !== z) { "Can't check equals transitivity: y:$y is same instance as z:$z" }
+        assumeTrue(x !== z) { "Can't check equals transitivity: x:$x is same instance as z:$z" }
+        assumeTrue(x.equals(y)) { "Can't check equals transitivity: x:$x isn't equal to y:$y" }
+        assumeTrue(y.equals(z)) { "Can't check equals transitivity: y:$y isn't equal to z:$z" }
+
+        // When
+        val result = x.equals(z)
+
+        // Then
+        assertThat(result).isTrue()
+    }
+
+    /**
+     * `equals` must be consistent.
+     * if a (relevant) field is changed, equals should detect it
+     */
+    @Test
+    fun `equals is consistent`(forge: Forge) {
+        // Given
+        val x = createInstance(forge)
+        val y = createUnequalInstance(x, forge) ?: return
+        assumeTrue(x !== y) { "Can't check equals consistency: x:$x is same instance as y:$y" }
+
+        // When
+        val result = x.equals(y)
+
+        // Then
+        assertThat(result).isFalse()
+    }
+
+    /**
+     * `equals` must be type sensitive.
+     * if the two compared instances are of different types, equals is false
+     */
+    @Test
+    fun `equals is type sensitive`(forge: Forge) {
+        // Given
+        val x = createInstance(forge)
+        val y = Object()
+        assumeTrue(x !== y) { "Can't check equals type sensitivity: x:$x is same instance as y:$y" }
+        assumeTrue(x.javaClass != y.javaClass) {
+            "Can't check equals transitivity: x:$x has same type as y:$y"
+        }
+
+        // When
+        val result = x.equals(y)
+
+        // Then
+        assertThat(result).isFalse()
+    }
+
+    // endregion
+
+    // region `hashCode` contract
+
+    /**
+     * `hashCode` must be consistent internaly.
+     * if a (relevant) field is changed, hashCode may detect it (but collisions can happen)
+     */
+    @Test
+    @Suppress("FunctionMaxLength")
+    fun `hashCode is internally consistent`(forge: Forge) {
+        // Given
+        val x = createInstance(forge)
+
+        var testedHashCode = 0
+        var differentHashCode = 0
+
+        // When
+        repeat(64) {
+            val y = createUnequalInstance(x, forge)
+            if (y != null && y !== x) {
```

Comment:
why do you need this comparison ? Can the `createUnequalInstace` function return the same instance ?